src/pybind/mgr/prometheus/metric.py

```diff
@@ -0,0 +1,102 @@
+import copy
+import math
+
+import static_metrics
+
+
+def get_metrics_spec():
+    """
+    Load the metrics configuration, and build a spec representing the set of
+    metrics we'll use.
+    """
+    metrics_spec = copy.deepcopy(static_metrics.groups)
+    for name, group in metrics_spec.items():
+        templates = group.pop('templates', dict())
+        defaults = group.pop('defaults', dict())
+        for name, item in group.get('metrics', dict()).items():
+            apply_defaults(defaults, item)
+            apply_templates(templates, name, item)
+            if 'name' not in item:
+                item['name'] = name
+    return metrics_spec
+
+
+def apply_defaults(defaults, item):
+    for name, value in defaults.items():
+        if name not in item:
+            item[name] = value
+    return item
+
+
+def apply_templates(templates, item_name, item):
+    for name, value in templates.items():
+        new_value = value.format(item_name)
+        if name not in item:
+            item[name] = new_value
+    return item
```

Comment:
This seems a bit complicated (first deepcopy, then alter the dict structure). Also the resulting structure has some redundant info in it. It looks like this: ```{group: "metrics": {shortname: {name: ..., description: ..., label: ..., type: ...}}}}```

Maybe something like this would be a bit more concise?
```
def get_metrics_spec():                                                        
    """                                                                        
    Load the metrics configuration, and build a spec representing the set of
    metrics we'll use.                                                         
    """                                                                        
    metrics_spec = {}                                                          
    for name, group in static_metrics.groups.items():                          
        metrics_spec[name] = {}                                                
        templates = group.get('templates', dict())                             
        defaults = group.get('defaults', dict())                               
        for m_name, item in group.get('metrics', dict()).items():              
            metrics_spec[name].update(apply_templates_and_defaults(templates,
                                                              defaults, m_name,
                                                              item))           
    return metrics_spec                                                        
                                                                               
                                                                               
def apply_templates_and_defaults(templates, defaults, item_name, item):    
    ret = {}                                                                   
    for name, value in templates.items():     
        ret[name] = value.format(item_name)                                                  
    for name, value in defaults.items():                                       
        ret[name] = value                                                      
    # overwrite items specific keys                                            
    ret.update(copy.deepcopy(item))                                            
    return {ret.pop('name', item_name): ret} 
```
This would result in a structure like ```{group: {name: {description: ..., label: ..., type: ...}}}```